tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+
+          val recElems = json.allFieldsOpt.get.toSeq.map { k =>
+            k -> typeDrivenBuild(json.getFieldOpt(k).get, fieldTypes(k))
+          }
+
+          // Records must contain at least 1 kv pair
+          if (recElems.isEmpty) throw new JsonDeserializationError(s"Records may not be empty.")
+
+          tla.rec(recElems: _*)
+
+        case TupT1(elems @ _*) =>
+          if (!json.allFieldsOpt.contains(Set(TUP_FIELD)))
+            throw new JsonDeserializationError(s"Tuple-type annotated values must declare a \"$TUP_FIELD\" field.")
+          val tupSeq = scalaFactory.asSeq(json.getFieldOpt(TUP_FIELD).get)
+
+          // The declared tuple length must match the # of elements
+          val valLen = tupSeq.length
+          val typeLen = elems.length
+          if (valLen != typeLen)
+            throw new JsonDeserializationError(
+                s"Tuple value and type have $valLen and $typeLen elements respectively."
+            )
+
+          val args = tupSeq.zip(elems).map { case (elemJson, elemT) =>
+            typeDrivenBuild(elemJson, elemT)
+          }
+          tla.tuple(args: _*)
+
+        case SetT1(elemT) =>
+          if (!json.allFieldsOpt.contains(Set(SET_FIELD)))
+            throw new JsonDeserializationError(s"Set-type annotated values must declare a \"$SET_FIELD\" field.")
+          val setSeq = scalaFactory.asSeq(json.getFieldOpt(SET_FIELD).get)
+          val args = setSeq.map { elemJson =>
+            typeDrivenBuild(elemJson, elemT)
+          }
+          // Empty collections have special builder ctors
+          args match {
+            case Seq() => tla.emptySet(elemT)
+            case seq   => tla.enumSet(seq: _*)
+          }
+
+        case FunT1(argT, resT) =>
+          if (!json.allFieldsOpt.contains(Set(MAP_FIELD)))
+            throw new JsonDeserializationError(s"Function-type annotated values must declare a \"$MAP_FIELD\" field.")
+
+          val funSeq = scalaFactory.asSeq(json.getFieldOpt(MAP_FIELD).get)
+
+          // Since ITF functions are enumerated, we must construct a funAsSet, instead of a funDef.
+
+          val args = funSeq.map { pairArrayJson =>
+            val pairArray = scalaFactory.asSeq(pairArrayJson)
+            if (pairArray.length != 2)
+              throw new JsonDeserializationError(
+                  s"Function entries must be 2-element arrays. Found entry with ${pairArray.length} elements.")
+            val Seq(kJson, vJson) = pairArray
+            val k = typeDrivenBuild(kJson, argT)
+            val v = typeDrivenBuild(vJson, resT)
+            tla.tuple(k, v) // for funAsSet, we need a set of paris
+          }
+          // Empty collections have special builder ctors
+          val set =
+            if (args.isEmpty) tla.emptySet(TupT1(argT, resT))
+            else tla.enumSet(args: _*)
+
+          // TODO: Check if keys are duplicated or let SetAsFun rule catch that? Aternatively,
+          // treat it as double-except semantics
```

Comment:
Seems this might be worth tracking in an issue?

Reply:
The thing is, I'm not 100% sure about  `SetAsFun` semantics, I was hoping for @konnov's input

Reply:
The semantics is here: https://github.com/informalsystems/apalache/pull/2273#discussion_r1018967098

Reply:
This just links to my comment btw.

Reply:
In short, `SetAsFun` would not do any error reporting on `SetAsFun({ <<1, 2>>, <<1, 3>>) })`. It would just interpret the set somehow. Looking at the code, it may even produce unsatisfiable constraints.